livre/lib/pipeline/HistogramFilter.cpp

```diff
@@ -0,0 +1,178 @@
+#include <livre/core/render/Frustum.h>
+
+namespace livre
+{
+
+namespace
+{
+const float infinite = 100000.0f;
+}
+
+struct HistogramFilter::Impl
+{
+    Impl( HistogramCache& histogramCache )
+        : _histogramCache( histogramCache )
+        , _dataSource( histogramCache.getDataCache().getDataSource( ))
+        , _components( _dataSource.getVolumeInfo().compCount )
+    {}
+
+    bool isCenterInViewport( const Frustum& frustum,
+                             const Boxf& worldBox,
+                             const Vector3f& minNdc = Vector3f( -1.0f ),
+                             const Vector3f& maxNdc = Vector3f( 1.0f )) const
+    {
+        Vector4f center = worldBox.getCenter();
+        center[ 3 ] = 1.0;
+
+        Vector4f mvpCenter = frustum.getMVPMatrix() * center;
+        Vector3f mvpCenterHom = mvpCenter / mvpCenter[ 3 ];
+
+        const Boxf ndcCube( minNdc, maxNdc );
+        return ndcCube.isIn( mvpCenterHom );
+    }
+
+    void scaleNdcIfBorder( const Viewport& border, Vector3f& minBox, Vector3f& maxBox ) const
+    {
+        const bool isNegXBorder = border.x() == 0.0f;
+        const bool isPosXBorder = border.x() + border.z() == 1.0f;
+        const bool isNegYBorder = border.y() == 0.0f;
+        const bool isPosYBorder = border.y() + border.w() == 1.0f;
+
+        if( isNegXBorder )
+            minBox[ 0 ] = -infinite;
+        if( isPosXBorder )
+            maxBox[ 0 ] = infinite;
+        if( isNegYBorder )
+            minBox[ 1 ] = -infinite;
+        if( isPosYBorder )
+            maxBox[ 1 ] = infinite;
+    }
+
+    void execute( const FutureMap& input, PromiseMap& output ) const
+    {
+        const auto frustum = input.get< Frustum >( "Frustum" ).front();
+        const auto viewport = input.get< Viewport >( "RelativeViewport" ).front();
+
+        Histogram histogramAccumulated;
+        for( const auto& cacheObjects: input.getFutures( "CacheObjects" ))
+            for( const auto& cacheObject: cacheObjects.get< ConstCacheObjects >( ))
+            {
+                const CacheId& cacheId = cacheObject->getId();
+                ConstCacheObjectPtr histCacheObject = _histogramCache.load( cacheId );
+                if( !histCacheObject->isLoaded( ))
+                    continue;
+
+                const ConstHistogramObjectPtr& histogramObj =
+                        std::static_pointer_cast< const HistogramObject >(
+                            histCacheObject );
+
+                // When a frame is rendered in multi-channel, multi-node, etc
+                // config, some nodes are rendered twice in sort-first renderings
+                // To avoid that, we check whether the center of the node is in
+                // frustum or, if node is in the viewport border. To do this we
+                // project the center of the node to the normalized device coordinates
+                // (NDC), in case of partial intersection we extend the borders of
+                // NDC in border directions
```

Comment:
// When a frame is rendered in multi-channel, multi-node, etc
// config, some nodes are rendered twice in sort-first renderings
// To avoid counting nodes twice, we check if the center of the node is in
// this frustum (because it can only be in one tile at a time). For viewports
// on the border of the absolute viewport, the frustum is virtually extended
// to infinity on the boundary.